cmd/notary/tuf.go

```diff
@@ -100,6 +111,79 @@ func (t *tufCommander) AddToCommand(cmd *cobra.Command) {
 	cmdTufRemove := cmdTufRemoveTemplate.ToCommand(t.tufRemove)
 	cmdTufRemove.Flags().StringSliceVarP(&t.roles, "roles", "r", nil, "Delegation roles to remove this target from")
 	cmd.AddCommand(cmdTufRemove)
+
+	cmdTufAddHash := cmdTufAddHashTemplate.ToCommand(t.tufAddByHash)
+	cmdTufAddHash.Flags().StringSliceVarP(&t.roles, "roles", "r", nil, "Delegation roles to add this target to")
+	cmdTufAddHash.Flags().StringVar(&t.sha256, notary.SHA256, "", "hex encoded sha256 of the target to add")
+	cmdTufAddHash.Flags().StringVar(&t.sha512, notary.SHA512, "", "hex encoded sha512 of the target to add")
+	cmd.AddCommand(cmdTufAddHash)
+}
+
+func (t *tufCommander) tufAddByHash(cmd *cobra.Command, args []string) error {
+	if len(args) < 3 || t.sha256 == "" && t.sha512 == "" {
+		cmd.Usage()
+		return fmt.Errorf("Must specify a GUN, target, byte size of target data, and at least one hash")
+	}
+	config, err := t.configGetter()
+	if err != nil {
+		return err
+	}
+
+	gun := args[0]
+	targetName := args[1]
+	targetSize := args[2]
+
+	targetInt64Len, err := strconv.ParseInt(targetSize, 0, 64)
+	if err != nil {
+		return err
+	}
+
+	// no online operations are performed by add so the transport argument
+	// should be nil
+	nRepo, err := notaryclient.NewNotaryRepository(
+		config.GetString("trust_dir"), gun, getRemoteTrustServer(config), nil, t.retriever)
+	if err != nil {
+		return err
+	}
+
+	targetHash := data.Hashes{}
+	if t.sha256 != "" {
+		if len(t.sha256) != notary.Sha256HexSize {
+			return fmt.Errorf("invalid sha256 hex contents provided")
+		}
+		sha256Hash, err := hex.DecodeString(t.sha256)
+		if err != nil {
+			return err
+		}
+		targetHash[notary.SHA256] = sha256Hash
+	}
+	if t.sha512 != "" {
+		if len(t.sha512) != notary.Sha512HexSize {
+			return fmt.Errorf("invalid sha512 hex contents provided")
+		}
+		sha512Hash, err := hex.DecodeString(t.sha512)
+		if err != nil {
+			return err
+		}
+		targetHash[notary.SHA512] = sha512Hash
+	}
+
+	// Manually construct the target with the given byte size and hashes
+	target := &notaryclient.Target{Name: targetName, Hashes: targetHash, Length: targetInt64Len}
+
+	// If roles is empty, we default to adding to targets
+	if err = nRepo.AddTarget(target, t.roles...); err != nil {
```

Comment:
Should check `len(targetHash) > 0`. Could add the check inside `AddTarget` to make it as widely applicable as possible.